data/read_data.rb

```diff
@@ -0,0 +1,37 @@
+def read_person_data
+  if File.exist?('./data/persons.json')
+
+    persons = JSON.parse(File.read('./data/persons.json'))
+    persons.map do |person|
+      if person['occupation'] == 'Teacher'
+        Teacher.new(person['age'], person['specialization'], person['name'])
+      else
+        Student.new(person['age'], person['classroom'], person['name'], person['parent_permission'])
+      end
+    end
+  else
+    []
+  end
+end
+
+def read_book
+  if File.exist?('./data/books.json')
+    books = JSON.parse(File.read('./data/books.json'))
+    books.map do |book|
+      Book.new(book['title'], book['author'])
+    end
+  else
+    []
+  end
+end
+
+def read_rentals
+  if File.exist?('./data/rentals.json')
+    rent = JSON.parse(File.read('./data/rentals.json'))
+    rent.map do |rental|
+      Rental.new(rental['date'], @person[rental['person_index']], @books[rental['book_index']])
+    end
+  else
+    []
+  end
+end
```

Comment:
- [ ] Kindly note that the requirements say to handle the case when one or more of the files are missing. In your case, if a file is not present then no data is preserved kindly fix it.

- [ ] Kindly note that you need to load these data when the app starts, actually you are just only creating instances of classes but they are not saved anywhere. All these methods should be executed in the `app.rb` so that when the application loads the preserved data should be loaded.


example :point_down: 

```rb
def read_person_data(person_json)
  if File.exist?('./data/persons.json')

    persons = JSON.parse(File.read('./data/persons.json'))
    persons.map do |person|
      if person['occupation'] == 'Teacher'
       person_json << Teacher.new(person['age'], person['specialization'], person['name']) 
      else
       person_json << Student.new(person['age'], person['classroom'], person['name'], person['parent_permission'])
      end
    end
  else
    []
  end
end
```

then in app.rb

```rb
require './data/read_data'



class App < Console
  def initialize
    super()
    @persons = []
    @books = []
    @rentals = []
  end
  def load_data
    read_person_data(@persons)
  end


.......... your content......................


end
```

and then in your main.rb

```.rb
require './app'

def main
  app = App.new
  app.load_data
  app.console_entry_point
end

main
```

### Kindly note that this code above :point_up:  is just an example you solve this the way you can, just make sure the preserved data is available when the app loads.

Reply:
✅ RESOLVED: We have ensured that the preserved data is loaded when the app starts.